docs/guide/ethereum-provider.md

```diff
@@ -112,83 +173,359 @@ params: [
   },
 ];
 
-ethereum.sendAsync(
-  {
+ethereum
+  .request({
     method: 'eth_sendTransaction',
-    params: params,
-    from: accounts[0], // Provide the user's account to use.
-  },
-  (err, response) => {
-    if (err) {
-      // Handle the error
-    } else {
-      // This always returns a JSON RPC response object.
-      // The result varies by method, per the RPC method specification
-      // For example, this method will return a transaction hash on success.
-      const result = response.result;
-    }
-  }
-);
+    params,
+  })
+  .then((result) => {
+    // The result varies by by RPC method.
+    // For example, this method will return a transaction hash hexadecimal string on success.
+  })
+  .catch((error) => {
+    // If the request fails, the Promise will reject with an error.
+  });
 ```
 
-### ethereum.sendAsync({ method: 'eth_requestAccounts'}, callback)
+## Events
 
-Requests to view the user's Ethereum address.
+The MetaMask provider implements the [Node.js `EventEmitter`](https://nodejs.org/api/events.html) API.
+This sections details the events emitted via that API.
+There are innumerable `EventEmitter` guides elsewhere, but you can listen for events like this:
 
-#### Example
+```javascript
+ethereum.on('accountsChanged', (accounts) => {
+  // Handle the new accounts, or lack thereof.
+  // "accounts" will always be an array, but it can be empty.
+});
+
+ethereum.on('chainChanged', (chainId) => {
+  // Handle the new chain.
+  // We recommend reloading the page.
+  window.location.reload();
+});
+```
+
+### connect
+
+```typescript
+interface ConnectInfo {
+  chainId: string;
+}
+
+ethereum.on('connect', handler: (connectInfo: ConnectInfo) => void);
+```
+
+The MetaMask provider emits this event when it first becomes able to submit RPC requests to a chain.
+In general, you can assume that the MetaMask provider is connected and has emitted this event by the time you are able to reference it.
+
+### disconnect
+
+```typescript
+ethereum.on('disconnect', handler: (error: ProviderRpcError) => void);
+```
+
+The MetaMask provider emits this event if it becomes unable to submit RPC requests to any chain.
+In general, this will only happen due to network connectivity issues or some unforeseen error.
+
+Once `disconnect` has been emitted, MetaMask will not accept any new requests until until `connect` is emitted, which may require reloading the page.
+
+### accountsChanged
+
+```typescript
+ethereum.on('accountsChanged', handler: (accounts: Array<string>) => void);
+```
+
+The MetaMask provider emits this event whenever the return value of the `eth_accounts` RPC method changes.
+`eth_accounts` returns an array that is either empty or contains a single account address that the user decided to expose to the requesting domain (in the case of a website, identified by the _hostname_ of its URL).
```

Comment:
I'm not sure this is correct exactly; we assign permissions by origin, not by hostname. We could probably say a little more about which account we're returning too (e.g. that it's the most recently used account).
Maybe something like this?
```suggestion
`eth_accounts` returns an array that is either empty or contains a single account address. The account address returned (if any) is the most recently used account that the requesting site has permission to access. Permissions are granted to _origins_, so all sites on the same origin share the same permissions.
```

Reply:
cf58561